### ISSUE TITLE
Julia crash on tiny change

### ISSUE THREAD

**Issue by bodono** [JuliaOpt/SCS.jl#31]
Changing three lines https://github.com/JuliaOpt/SCS.jl/blob/master/test/options.jl#L45 from
```
45 primal_sol = m.primal_sol
46 dual_sol = m.dual_sol
47 slack = m.slack
```
to
```
45 primal_sol = deepcopy(m.primal_sol)
46 dual_sol = deepcopy(m.dual_sol)
47 slack = deepcopy(m.slack)
```

results in 
```
julia(32256,0x7fff7b9f6300) malloc: *** error for object 0x7faee77bb0a0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
```

my guess is that this is in julia itself, but I can't tell, could also be in MathProgBase or SCS.jl. @mlubin do you have any insight on where this could be coming from?

**Comment by bodono** [comment]
I should mention I'm on OSX 10.10.3, 64 bit, using the standard homebrew SCS 1.1.5.

**Comment by mlubin** [comment]
@bodono, does SCS ever call ``free()`` on memory passed to it?

**Comment by mlubin** [comment]
Where did the memory [here](https://github.com/bodono/SCS.jl/blob/3d86eef52a73ae163d1f046850ee7fa462d4d553/src/high_level_wrapper.jl#L141) come from? Why is it converted to a pointer and then back to an array?

**Comment by bodono** [comment]
It never calls free on memory passed in.

I don't know where that memory came from. Looks like it used to be
```
    x = copy(pointer_to_array(ptr_x, n))
    y = copy(pointer_to_array(ptr_y, m))
    s = copy(pointer_to_array(ptr_s, m))
```

should we change it back to that?

**Comment by mlubin** [comment]
Does SCS allocate that memory? Is there a way to free it from inside SCS?

**Comment by mlubin** [comment]
It's definitely safer to copy the memory into a julia array instead of asking julia to take over the garbage collection for memory which it didn't allocate.

**Comment by bodono** [comment]
Yes, in this case it is allocated inside of SCS, because SCS checks if the memory is there before and if not then allocates it.

I can change it back then in another PR. Why does julia allow that operation if it's unsafe then?

**Comment by bodono** [comment]
But if that memory is not cleared out by julia then it will result in a memory leak.

**Comment by mlubin** [comment]
Can we just allocate the memory in julia from the beginning?

> Why does julia allow that operation if it's unsafe then?

Well, the gloves are off when you're using pointers. There are some issues on various platforms with freeing memory using a different runtime library than where it was allocated, but julia doesn't stop you from trying.

**Comment by bodono** [comment]
Sounds good, I'll add it to https://github.com/JuliaOpt/SCS.jl/pull/32 once I've made the change.

**Comment by madeleineudell** [comment]
closed by #32
